server/pulp/plugins/file/model_distributor.py

```diff
@@ -0,0 +1,249 @@
+from gettext import gettext as _
+import csv
+import errno
+import logging
+import os
+import shutil
+import traceback
+
+from pulp.common.plugins.distributor_constants import MANIFEST_FILENAME
+from pulp.plugins.distributor import Distributor
+from pulp.plugins.file.distributor import FilePublishProgressReport
+from pulp.server.controllers import repository as repo_controller
+from pulp.server.db import model
+from pulp.server.managers.repo import _common as common_utils
+
+
+BUILD_DIRNAME = 'build'
+
+_logger = logging.getLogger(__name__)
+
+
+class FileDistributor(Distributor):
+    """
+    Distribute files on the filesystem.
+    """
+
+    def __init__(self):
+        super(FileDistributor, self).__init__()
+        # Initialize instance variables used for writing out the metadata
+        self.metadata_file = None
+        self.metadata_csv_writer = None
+
+    @classmethod
+    def metadata(cls):
+        """
+        Advertise the capabilities of the mighty FileDistributor.
+
+        :return: The description of the impressive FileDistributor's capabilities.
+        :rtype:  dict
+        """
+        raise NotImplementedError()
+
+    def distributor_removed(self, repo, config):
+        """
+        Delete the published files from our filesystem
+
+        Please also see the superclass method definition for more documentation on this method.
+
+        :param repo: metadata describing the repository
+        :type  repo: pulp.server.db.model.Repository
+
+        :param config: plugin configuration
+        :type  config: pulp.plugins.config.PluginCallConfiguration
+        """
+        self.unpublish_repo(repo, config)
+
+    def publish_repo(self, repo, publish_conduit, config):
+        """
+        Publish the repository.
+
+        :param repo:            metadata describing the repo
+        :type  repo:            pulp.server.db.model.Repository
+        :param publish_conduit: The conduit for publishing a repo
+        :type  publish_conduit: pulp.plugins.conduits.repo_publish.RepoPublishConduit
+        :param config:          plugin configuration
+        :type  config:          pulp.plugins.config.PluginConfiguration
+        :param config_conduit: Configuration Conduit;
+        :type config_conduit: pulp.plugins.conduits.repo_validate.RepoConfigConduit
+        :return:                report describing the publish operation
+        :rtype:                 pulp.plugins.model.PublishReport
+        """
+        progress_report = FilePublishProgressReport(publish_conduit)
+        _logger.info(_('Beginning publish for repository <%(repo)s>') % {'repo': repo.id})
+
+        try:
+            progress_report.state = progress_report.STATE_IN_PROGRESS
+            repo_model = model.Repository.objects.get_repo_or_missing_resource(repo.id)
```

Comment:
If the argument `repo` ends up as type `pulp.server.db.model.Repository` then this query can be removed since it returns that type already (I believe).